sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -52,19 +52,27 @@ case class DataSourceV2Relation(
     case c: FunctionCatalog => c
   }
 
-  override lazy val metadataOutput: Seq[AttributeReference] = table match {
-    case hasMeta: SupportsMetadataColumns =>
-      val resolve = conf.resolver
-      val outputNames = outputSet.map(_.name)
-      def isOutputColumn(col: MetadataColumn): Boolean = {
-        outputNames.exists(name => resolve(col.name, name))
+  override lazy val metadataOutput: Seq[AttributeReference] = {
+    val metadataColFromOutput = output.filter(_.isMetadataCol)
```

Comment:
to be safe, let's only do it under `case hasMeta: SupportsMetadataColumns`

Reply:
also add some comments to explain when it happens

Reply:
addressed